### DISCUSSION TITLE
Bug fix. There are 2 missing yaw angles in SR optimization

### DISCUSSION THREAD

**Discussion by dhcho347** [NatLabRockies/floris#539] [Q&A]
Hi, I think I bring back the 2 yaw angles in SR optimization.

with default input parameter
 - Ny_passes = [5,4]
 - minimum_yaw_angle=0.0, maximum_yaw_angle=25.0

in the second pass, we lost 2 angles due to the clip

The length of the first and last subdivision has changed (from 6.25 to 3.125)

This makes the center does not match to the first pass no longer.
https://docs.google.com/spreadsheets/d/1NNB-2VbZFfed8HEw6xYvsuvMsQ1bM-K3LYQDsSkoFlA/edit#gid=1989138220

![image](https://user-images.githubusercontent.com/70368625/202940577-0d7185f6-11d0-4e3b-ade1-e6a1a475be2f.png)

So, we can't get the 23.4375 degree result from ex10.
(The others (21.875, 22.65625, 24.21875, 25) are there)

We need to change the values in first and last subdivision.
for example, last subdivision,
We have to take 23.4375 instead of 25

#Enhancement
As one of the temporary solutions,
I've add 5 lines in the method _generate_evaluation_grid(self, pass_depth, turbine_depth):

https://github.com/dhcho347/floris/commit/234f88391d40b4fae828413a0cff9400c34424f4#

#Results
After change, I got a 23.4375 degree result from ex10.

Before, yaw_angles_subset in last division was :  [21.875, 22.6563, (x 23.4375), 24.2188, 25] 
after, yaw_angles_subset in last division was :  [21.875, 22.6563, 23.4375, 24.2188, (x 25 in first pass)] 

![image](https://user-images.githubusercontent.com/70368625/202938745-9637cd57-a832-4184-853b-69b933ff5840.png)

Please, check this one 
Thank you

**Comment by clun1** [comment]
I'm curious how did you get this table and where did the output come from？
thank advance

**Reply by dhcho347** [reply]
Hi, @clun1 
I just follow the code in yaw_optimizer_sr.py
It's relating with https://github.com/NREL/floris/blob/develop/floris/tools/optimization/yaw_optimization/yaw_optimizer_sr.py#L159
```python
 def _generate_evaluation_grid(self, pass_depth, turbine_depth):
```

I just made the table with below codes
https://github.com/NREL/floris/blob/develop/floris/tools/optimization/yaw_optimization/yaw_optimizer_sr.py#L171-L207

This is similar to binary search algorithm. but use 4 ,etc... instead of 2 for dividing section
```python
        # Initialize yaw angles to evaluate, 'Ny' times the wind rose
        Ny = self.Ny_passes[pass_depth]
        evaluation_grid = np.tile(self._yaw_angles_opt_subset, (Ny, 1, 1, 1))

        # Get a list of the turbines in order of x and sort front to back
        for iw in range(self._nwinddirections_subset):
            turbid = self.turbines_ordered_array_subset[iw, turbine_depth]  # Turbine to manipulate

            # Grab yaw bounds from self
            yaw_lb = self._yaw_lbs[iw, :, turbid]
            yaw_ub = self._yaw_ubs[iw, :, turbid]
```
 clip both ub and lb with input range limit
```python
            # Saturate to allowable yaw limits
            yaw_lb = np.clip(
                yaw_lb,
                self.minimum_yaw_angle[iw, :, turbid],
                self.maximum_yaw_angle[iw, :, turbid]
            )
            yaw_ub = np.clip(
                yaw_ub,
                self.minimum_yaw_angle[iw, :, turbid],
                self.maximum_yaw_angle[iw, :, turbid]
            )
```
dividing section with input number.
with Ny_passes = [5,4] input, first and second sections are divided by 4 (4 section, 5 points)
  - first pass, lb=0, ub=5 => 5 discrete pts = [0, 6.25, 12,5 , 18.75, 25]
 
but in the second pass, we drop a center point and just calculate 4 points
  - second pass, 
     v 1st, lb=0, ub=6.25 => 5 discrete pts = [0, 0.781, 1.562(middle) , 2.343, 3.125]
     v 2nd,  lb=6.25, ub=12.5 => 5 discrete pts = [3.125, 4.6875, 6.25(middle) , 7.812, 9.375]
     v 3rd, ....

```python
            if pass_depth == 0:
                # Evaluate all possible coordinates
                yaw_angles_subset = np.linspace(yaw_lb, yaw_ub, Ny)
            else:
                # Remove middle point: was evaluated in previous iteration
                c = int(Ny / 2)  # Central point (to remove)
                ids = [*list(range(0, c)), *list(range(c + 1, Ny + 1))]
                yaw_angles_subset = np.linspace(yaw_lb, yaw_ub, Ny + 1)[ids]

            evaluation_grid[:, iw, :, turbid] = 
```

I hope this has been answered.

**Comment by paulf81** [comment]
@dhcho347 I apologize that I overlooked this discussion earlier.  There have been some important bugfixes to SR made, but I'm not sure whether this resolves the issue you describe above.  Would you mind to post a minimal script that reproduces the issue we can use to create an issue and investigate?

**Reply by dhcho347** [reply]
Hi, @paulf81 
I tested just update with #590. 
The problem still remains. Because the causes are different 
(it doesn't relate with fluctuation. 
It relate with clip and remove middle point https://github.com/NREL/floris/discussions/539#discussioncomment-5669098 )

Here is test result
![image](https://user-images.githubusercontent.com/70368625/233266181-1defda7f-2c38-4271-831c-4c0a08c3b124.png)



This is a temporary code I'm using: https://github.com/dhcho347/floris/commit/234f88391d40b4fae828413a0cff9400c34424f4

https://github.com/dhcho347/floris/commit/234f88391d40b4fae828413a0cff9400c34424f4#diff-cda3b9c49243054631ba95f53374bf1805dc57b1105e656d08776aa9b1228331R167-R169
```python
        if 1: #dh. # yaw angles are evaluated in first iteration
            lb=self._minimum_yaw_angle_subset[0,0,0]; ub=self._maximum_yaw_angle_subset[0,0,0]
            yaw_in_first_pass = np.linspace([lb],[ub],self.Ny_passes[0])
```


https://github.com/dhcho347/floris/commit/234f88391d40b4fae828413a0cff9400c34424f4#diff-cda3b9c49243054631ba95f53374bf1805dc57b1105e656d08776aa9b1228331R203-R210
```python
                if 1: 
                    yaw_angles_subset = np.linspace(yaw_lb, yaw_ub, Ny + 1)
                    if (yaw_angles_subset == yaw_in_first_pass).all(): yaw_angles_subset=yaw_angles_subset[1:]
                    else : yaw_angles_subset = np.array([i for i in yaw_angles_subset if i not in yaw_in_first_pass]) #dh
                else: 
                    c = int(Ny / 2)  # Central point (to remove)
                    ids = [*list(range(0, c)), *list(range(c + 1, Ny + 1))]  
                    yaw_angles_subset = np.linspace(yaw_lb, yaw_ub, Ny + 1)[ids]
```
This line should be more generalize
```python
if (yaw_angles_subset == yaw_in_first_pass).all(): yaw_angles_subset=yaw_angles_subset[1:]
```

I'm sharing files so you can test it.
https://github.com/dhcho347/dh_test/commit/7677632f8d5e9e7a89b97cd2f4b880ad054b005a
  1. input file: cc_100mw.yaml (WTG 20ea, 100MW wind farm)
  2. run script: 2run_Missing angles_sweep_speed_directions.py


 Thank you

**Reply by dhcho347** [reply]
I made my temporary code more general. https://github.com/dhcho347/floris/commit/9048d46b886d4935f19a8c04d19c12bdd3491cad
